python-squid/src/main/java/org/sonar/python/metrics/MetricsVisitor.java

```diff
@@ -0,0 +1,170 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.python.metrics;
+
+import com.intellij.psi.PsiComment;
+import com.intellij.psi.PsiElement;
+import com.intellij.psi.PsiWhiteSpace;
+import com.intellij.psi.impl.source.tree.LeafPsiElement;
+import com.jetbrains.python.PyTokenTypes;
+import com.jetbrains.python.psi.PyExceptPart;
+import com.jetbrains.python.psi.PyExpression;
+import com.jetbrains.python.psi.PyExpressionStatement;
+import com.jetbrains.python.psi.PyFile;
+import com.jetbrains.python.psi.PyIfPart;
+import com.jetbrains.python.psi.PyIfStatement;
+import com.jetbrains.python.psi.PyRecursiveElementVisitor;
+import com.jetbrains.python.psi.PyStatement;
+import com.jetbrains.python.psi.PyStringLiteralExpression;
+import com.jetbrains.python.psi.PyTryExceptStatement;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+import org.sonar.python.frontend.PythonTokenLocation;
+
+public class MetricsVisitor extends PyRecursiveElementVisitor {
+
+  private final boolean ignoreHeaderComments;
+  private Set<Integer> linesOfCode = new HashSet<>();
+  private Set<Integer> executableLines = new HashSet<>();
+  private Set<Integer> linesOfComments = new HashSet<>();
+  private Set<Integer> noSonar = new HashSet<>();
+  private boolean firstNonCommentSeen = false;
+
+  public MetricsVisitor(boolean ignoreHeaderComments) {
+    this.ignoreHeaderComments = ignoreHeaderComments;
+  }
+
+  @Override
+  public void visitElement(PsiElement element) {
+    // track the first meaningful element which is not a comment to handle ignoreHeaderComments
+    if (!((element instanceof PyFile) || (element instanceof PsiComment) || (element instanceof PsiWhiteSpace))) {
+      firstNonCommentSeen = true;
+    }
+
+    // track lines of code
+    if (element instanceof LeafPsiElement
+      && !(element instanceof PsiWhiteSpace)
+      && !(element instanceof PsiComment)
+      && element.getNode().getElementType() != PyTokenTypes.DOCSTRING) {
+      PythonTokenLocation location = new PythonTokenLocation(element);
+      for (int line = location.startLine(); line <= location.endLine(); line++) {
+        linesOfCode.add(line);
+      }
+    }
+
+    // track executable lines of code
+    if (element instanceof PyStatement) {
+      handlePyStatement(element);
+    }
+
+    // track lines of comments and no_sonar lines
+    if (element instanceof PsiComment) {
+      if (ignoreHeaderComments && !firstNonCommentSeen) {
+        firstNonCommentSeen = true;
```

Comment:
I believe this is wrong for any file which has at least 2 lines of comment at the beginning.

Reply:
@pynicolas indeed, https://github.com/SonarSource/sonar-python/pull/254/commits/955f7b8cadd4c7b1ad8e4ff545721d1384713d17.